sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DoNotCheckZeroSizeCollection.cs

```diff
@@ -0,0 +1,63 @@
+/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2017 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.VisualBasic;
+using Microsoft.CodeAnalysis.VisualBasic.Syntax;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+using SonarAnalyzer.Helpers.VisualBasic;
+
+namespace SonarAnalyzer.Rules.VisualBasic
+{
+    [DiagnosticAnalyzer(LanguageNames.VisualBasic)]
+    [Rule(DiagnosticId)]
+    public sealed class DoNotCheckZeroSizeCollection : DoNotCheckZeroSizeCollectionBase<SyntaxKind, BinaryExpressionSyntax, ExpressionSyntax>
+    {
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+        protected override DiagnosticDescriptor Rule => rule;
+        protected sealed override Helpers.GeneratedCodeRecognizer GeneratedCodeRecognizer => Helpers.VisualBasic.GeneratedCodeRecognizer.Instance;
+        protected override SyntaxKind GreaterThanOrEqualExpression => SyntaxKind.GreaterThanOrEqualExpression;
+        protected override SyntaxKind LessThanOrEqualExpression => SyntaxKind.LessThanOrEqualExpression;
+        protected override ExpressionSyntax GetLeftNode(BinaryExpressionSyntax binaryExpression) => binaryExpression.Left;
+        protected override ExpressionSyntax GetRightNode(BinaryExpressionSyntax binaryExpression) => binaryExpression.Right;
+        protected override ExpressionSyntax RemoveParentheses(ExpressionSyntax expression) => expression.RemoveParentheses();
+        protected override string IEnumerableTString { get; } = "IEnumerable(Of T)";
+
+        protected override ISymbol GetSymbol(SyntaxNodeAnalysisContext context, ExpressionSyntax expression)
+        {
+            while (true)
```

Comment:
I strongly recommend to rewrite this `while`. I really don't like to see a `while (true)` when you do have an exiting path.

Reply:
This is the cleanest way to express it I found. Do you have a suggestion on what to change it to?

Reply:
@Evangelink can you write this with Linq only? :) I am not sure if it will be better (even with Aggregate)

Reply:
Cannot do Linq in here but at some point it might be worth adding an extension method for this walking up ancestors as this bit of code happens quite a lot.

Reply:
+1 for the extension method idea

This is very nice, but won't work:
- conditionalAccess.WhenNotNull would throw for non-conditional access
- the loop terminates when expression is null - so GetSymbolInfo is guaranteed to fail

These are fixable, but I think the version with while-true is cleaner than this:
```
ExpressionSyntax previousExpression, nextExpression = expression;
do
{
    previousExpression = RemoveParentheses(nextExpression);
    var conditionalAccess = expression as ConditionalAccessExpressionSyntax;
    nextExpression = conditionalAccess?.WhenNotNull;
}
while (nextExpression != null);
```

Reply:
Yeah I went a bit too fast here. The pattern used in a couple of other classes (not saying is the best) is:
```
var currentExpression = expression;
var conditionalAccess = RemoveParentheses(currentExpression) as ConditionalAccessExpressionSyntax;
while (conditionalAccess != null)
{
    currentExpression = conditionalAccess.WhenNotNull;
    conditionalAccess = RemoveParentheses(currentExpression) as ConditionalAccessExpressionSyntax;
}
```

Reply:
Yeah, that's the one I don't like as it's more complex. Hope one day we'll find the version that satisfies us both.